#!/usr/bin/env python

import sys, os
from Bio import SeqIO
import argparse

parser = argparse.ArgumentParser(description='Script to iteratively assemble targeted genes in a related species using RNA-seq data')
parser.add_argument('-t', dest = 'TargetFile', type = str, required=True,  help = 'Path to tab-delimited file with the names of the gene separated by path to the fasta file. Generated by split_multifasta_by_gene.py')
parser.add_argument('-1', dest= 'Left', type = str, required=True, help ='Left reads file in fastq format')
parser.add_argument('-2', dest= 'Right', type = str, required=True, help ='Right reads file in fastq format')
parser.add_argument('-i', dest= 'maxIter', type = int, default= '5', help ='Maximum number of iterations. Min=3. Default = 5')
parser.add_argument('-SS_lib_type', dest='SS', type = str, choices=['FR', 'RF'], default = '', help ='Stranded specific library type for Trinity (RF or FR). Default=Blank')
parser.add_argument('-p', dest= 'Threads', type = int, default= 16, help ='Number of threads to use. Default = 16')
args = parser.parse_args()


def run_trinity_first(left_reads, right_reads, target):
	cmd = 'Trinity --max_memory 50G --seqType fq --left %s --right %s %s --min_contig_length 100 --CPU %s --output temp/%s.trinity --full_cleanup &> /dev/null' % (left_reads, right_reads, '--SS_lib_type ' +args.SS,str(args.Threads), target)
	os.system(cmd)

def run_trinity_next(left_reads, right_reads, target):
	cmd = 'Trinity --max_memory 50G --seqType fq --left %s --right %s %s --min_contig_length 200 --CPU %s --output temp/%s.trinity --full_cleanup &> /dev/null' % (left_reads, right_reads, '--SS_lib_type ' +args.SS, str(args.Threads), target)
	os.system(cmd)

def run_bowtie2_idx(target_fasta, target_name):
	cmd = 'bowtie2-build -q %s temp/%s &> /dev/null' % (target_fasta, target_name)
	os.system(cmd)

def run_bowtie2_first(left_reads, target):
	cmd = 'bowtie2 -p %s --local --gbar 1 --mp 4 -D 20 -R 3 -N 1 -L 20 -i S,1,0.50 -x temp/%s -U %s --al temp/mapped_left.fastq -S temp/alignments.sam &> /dev/null'  % (str(args.Threads), target, left_reads)
	os.system(cmd)

def run_bowtie2_next(left_reads, target):
	cmd = 'bowtie2 -p %s --very-fast-local -x temp/%s -U %s --al temp/mapped_left.fastq -S temp/alignments.sam &> /dev/null'  % (str(args.Threads), target, left_reads)
	os.system(cmd)
	
def best_blast(gene_name, Trinity_fasta, target_fasta):
	make_blast_db = 'makeblastdb -in %s -dbtype nucl -parse_seqids -out temp/blastdb &> /dev/null' % (Trinity_fasta)
	os.system(make_blast_db)
	blast_cmd = 'blastn -db temp/blastdb -query %s -out temp/blastout.txt -outfmt 6' % (target_fasta)
	os.system(blast_cmd)
	firstline = open('temp/blastout.txt').readline()
	tophit = firstline.split('\t')[1]
	gene = '_'.join(tophit.split('_')[:-1])
	output = open('temp/%s.besthit.fasta'%gene_name, 'w')
	for seq_record in SeqIO.parse(Trinity_fasta, 'fasta'): #Trinity_fasta
		if gene in seq_record.id:
			SeqIO.write(seq_record, output, 'fasta')

def get_mates(mapped_left, right_idx):
	count = 0
	output = open('temp/right_mates.fastq', 'w')
	left = SeqIO.parse(mapped_left, 'fastq')
	for seq in left:
		count += 1
		if seq.id.endswith('/1'):
			right_id = seq.id.replace('/1','/2')
			SeqIO.write(right_idx[right_id], output, 'fastq')
		else:
			right_id = seq.id.replace(' 1:',' 2:')
			SeqIO.write(right_idx[right_id], output, 'fastq')
	return count
	
#target_list, left_reads, right_reads, iterations = sys.argv[1:]
print 'Indexing reads...'
right_idx = SeqIO.index(args.Right, 'fastq')
if not os.path.isdir('./temp/'):
	os.mkdir('./temp/')
with open(args.TargetFile, 'rU') as t:
	for line in t:
		target_gene = line.split('\t')[0]
		target_fasta = line.split('\t')[1].strip('\n')
		rcount0 = 0
		try:
			for i in range(0,int(args.maxIter)):
				if i == 0:
					# loose bowtie and trinity
					print '%s Iteration 0'%target_fasta
					run_bowtie2_idx(target_fasta, target_gene)
					print 'Aligning left reads to target'
					run_bowtie2_first(args.Left, target_gene)
					print 'Getting mate pairs'
					get_mates('temp/mapped_left.fastq', right_idx)
					print 'Assembling with Trinity'
					run_trinity_first('temp/mapped_left.fastq', 'temp/right_mates.fastq', target_gene)
				if i == 1:
					# map to all trinity contigs, more strict alignment and get best trinity contig
					print 'Iteration 1'
					run_bowtie2_idx('temp/%s.trinity.Trinity.fasta'%target_gene, target_gene)
					print 'Aligning left reads to new targets'
					run_bowtie2_next(args.Left, target_gene)
					print 'Getting mate pairs'
					get_mates('temp/mapped_left.fastq', right_idx)
					print 'Assembling with Trinity'
					run_trinity_next('temp/mapped_left.fastq', 'temp/right_mates.fastq', target_gene)
					print 'Finding best contigs'
					best_blast(target_gene,'temp/%s.trinity.Trinity.fasta'%target_gene, target_fasta)
				if i > 1:
					# now just use best contigs for mapping
					print 'Iteration %s' % str(i)
					run_bowtie2_idx('temp/%s.besthit.fasta' % (target_gene), target_gene)
					print 'Aligning left reads to new targets'
					run_bowtie2_next(args.Left, target_gene)
					print 'Getting mate pairs'
					rcount = get_mates('temp/mapped_left.fastq', right_idx)
					print str(rcount)+' reads mapped'
					if rcount > rcount0:
						rcount0 = rcount
						print 'Assembling with Trinity'
						run_trinity_next('temp/mapped_left.fastq', 'temp/right_mates.fastq', target_gene)
						print 'Finding best contigs'
						best_blast(target_gene,'temp/%s.trinity.Trinity.fasta'%target_gene, target_fasta)
					else:
						print 'Did not map any more reads this iteration'
						break
			os.rename('temp/%s.trinity.Trinity.fasta' % target_gene, '%s.trinity.Trinity.fasta' % target_gene)
			os.rename('temp/%s.besthit.fasta' % target_gene, '%s.besthit.fasta' % target_gene)
		except:
			print 'Found no contigs/ blast hits'
			try:
				os.rename('temp/%s.trinity.Trinity.fasta' % target_gene, '%s.trinity.Trinity.fasta' % target_gene)
				os.rename('temp/%s.besthit.fasta' % target_gene, '%s.besthit.fasta' % target_gene)
				continue
			except:
				continue
